Add Presto compatible hasher.

kgpai opened the pull request:
I created this PR since #738 would grow too long with this change. 

* Compatible with Presto's Type.hash()
* Ensures that values produced are equivalent to those created by Presto. 

Note: A subsequent PR will add an e2e test with Presto using checksum to ensure compatibility.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,217 @@
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DATE>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*_vector.get(), row);
```
Does this work because Date struct has a single int32_t field? This is hacky though.

kgpai replied:
Yes , do you recommend another way in which I can get the raw int out for hash purposes ?

mbasmanova replied:
I'd use `vector_->valueAt<Date>(row).days()`.

kgpai commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,217 @@
+  const charPtr buf = reinterpret_cast<const charPtr>(&value);
+  int64_t u;
+  charPtr pu = reinterpret_cast<charPtr>(&u);
+  std::copy(buf, buf + sizeof(double), pu);
```
No, its the same - fixing.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DOUBLE>(vector_size_t row) {
+  auto value = vector_->valueAt<double>(row);
+  int64_t u = *reinterpret_cast<int64_t*>(&value);
```
- DecodedVector::valueAt<T> already does reinterpret_cast, hence, you can simplify this code to 
 
`vector->valueAt<int64_t>(row)`

- Renaming getHashForIntTypes and getHashForVarCharTypes to hashInteger and hashStringView might improve readability:

```
return hashInteger(vector->valueAt<int64_t>(row));
...
return hashStringView(vector->valueAt<StringView>(row));
```

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+
+namespace {
+
+int64_t rotateLeft(const int64_t num, const int32_t distance) {
```
This function is used only once in a method which itself is a one-liners. I suggest to fold it into the single call site.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t getHashForIntTypes(const T& value) {
+  return rotateLeft(value * 0xC2B2AE3D27D4EB4FL, 31) * 0x9E3779B185EBCA87L;
```
Where do these constants come from? Would be nice to give them names and add a comment to document their origin.

kgpai replied:
These constants come from Presto. Not sure what to name them, as there arent names there either, but I will try.

mbasmanova replied:
Do these come from https://github.com/airlift/slice/blob/master/src/main/java/io/airlift/slice/XxHash64.java#L27 ?

I do see them being named PRIME64_1, PRIME64_2,.. PRIME64_5.

mbasmanova replied:
I also see these constants in velox/external/xxhash.h . Can't we just use some function from there instead of writing our own?

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+
+  if (children_.empty()) {
+    children_.reserve(1);
+    SelectivityVector elementRows(baseArray->elements()->size());
+    auto elements =
+        std::make_shared<DecodedVector>(*baseArray->elements(), elementRows);
+    children_.push_back(std::make_unique<PrestoHasher>(elements));
```
This code repeats multiple times. Consider extracting a helper method to adding a constructor to PrestoHasher that takes a vector.

```
children_.push_back(makePrestoHasher(baseArray->elements()));
```

or

```
children_.push_back(std::make_unique<PrestoHasher>(baseArray->elements());
```

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::ARRAY>(vector_size_t row) {
+  int64_t hash = 0;
```
nit: move this variable closer to first use for readability

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = vector_->indices();
+
+  auto size = rawSizes[indices[row]];
```
rawSizes and rawOffsets are used only once, hence, it would be simpler to use sizeAt and offsetAt methods:

```
auto index = vector->indexOf(row);
auto size =  baseArray->sizeAt(index);
auto offset =  baseArray->offsetAt(index);
```

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+  auto size = rawSizes[indices[row]];
+  auto offset = rawOffsets[indices[row]];
+
+  if (children_.empty()) {
```
Consider moving this logic into constructor to simplify the loop over rows.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+  auto valueIndices = children_[1]->indices();
+
+  for (int i = 0; i < size; i++) {
+    auto keyHash = children_[0]->computeHash(keyIndices[offset] + i);
```
Why not `children_[0]->computeHash(offset + i);`?

kgpai replied:
If key is a dictionary vector etc, wont I need to look it up in keyIndices , will the offset + i work for every type of vector ?

mbasmanova replied:
If that's the case, won't you need to do the same for ARRAY? Perhaps, add some tests for these scenarios.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,208 @@
+int64_t PrestoHasher::computeHash<TypeKind::ROW>(vector_size_t row) {
+  int64_t hash = 1;
+  auto baseRow = vector_->base()->as<RowVector>();
+  bool isPrimitive = true;
```
This variable is not used. Remove.

mbasmanova reviewed (commented):
@kgpai Some further comments.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  auto kind = vector_->base()->typeKind();
+  return VELOX_DYNAMIC_TYPE_DISPATCH(computeHash, kind, row);
```
Type dispatch for every row tends to be slow. Consider adding a benchmark to compare hashing a flat vector of integers using PrestoHasher and hand-written loop.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,182 @@
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TIMESTAMP>(vector_size_t row) {
+  return hashInteger<int64_t>(*vector_.get(), row);
+}
```
Timestamp is a struct with 2 fields. It doesn't seem valid to cast it to int64_t. I wonder if the right logic is to call Timestamp::toMillis().

```
  int64_t seconds_;
  uint64_t nanos_;
```

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,182 @@
+FOLLY_ALWAYS_INLINE int64_t
+hashStringView(const DecodedVector& vector, vector_size_t row) {
+  auto input = vector.valueAt<StringView>(row);
+  int64_t hash = XXH64(input.data(), input.size(), 0);
```
nit: this variable is not needed

```
return XXH64(input.data(), input.size(), 0);
```

kgpai commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t hashInteger(const T& value) {
+  int32_t distance = 31;
```
Note : making this constant breaks hash calculation.

mbasmanova commented:
@kgpai macos-build is failing. Would you take a look?

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+}
+
+FOLLY_ALWAYS_INLINE int64_t
+hashStringView(const DecodedVector& vector, vector_size_t row) {
```
I'd expect this function to be similar to hashInteger, but hashInteger checks for null while this function doesn't. Why is it not necessary for this function to check for null?

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+} // namespace
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::BIGINT>(
```
All the integer types can use a single implementation:

```
template <TypeKind kind>
void PrestoHasher::bulkHash(
    const SelectivityVector& rows,
    BufferPtr& buffer) {
  using T = typename TypeTraits<kind>::NativeType;
  bulkHashIntegral<T>(*vector_.get(), rows, buffer);
}
```

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashFn(rows, buffer, [&](auto row) -> int64_t {
+    return hashInteger(vector_->valueAt<Date>(row).days());
```
Is it not necessary to check for nulls?

kgpai replied:
Adding these checks..and test cases..

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+void PrestoHasher::bulkHash<TypeKind::ARRAY>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  auto baseArray = vector_->base()->as<velox::ArrayVector>();
```
drop velox:: prefix here and in other places

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+    BufferPtr& buffer) {
+  auto baseArray = vector_->base()->as<velox::ArrayVector>();
+
+  auto convertedRows = velox::functions::toElementRows(
```
naming: elementRows

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+  auto convertedRows = velox::functions::toElementRows(
+      baseArray->elements()->size(), rows, baseArray, vector_->indices());
+
+  BufferPtr childBuffer =
```
naming: elementHashes

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto rawNulls = baseArray->rawNulls();
+  auto rawChildHashes = childBuffer->as<int64_t>();
```
naming: rawElementHashes

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+  auto rawOffsets = baseArray->rawOffsets();
+  auto rawNulls = baseArray->rawNulls();
+  auto rawChildHashes = childBuffer->as<int64_t>();
+  auto hashBuffer = buffer->asMutable<int64_t>();
```
naming: rawHashes

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+
+  rows.applyToSelected([&](auto row) {
+    int64_t hash = 0;
+    if (!(rawNulls && bits::isBitNull(rawNulls, row))) {
```
row -> indices[row] ? 

It would be nice to add test coverage to catch this type of bugs.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+    hashBuffer[row] = hash;
+  });
+
+  // Free the buffer ?
```
This is redundant. The buffer will get freed once the reference count goes down to 0.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::MAP>(
```
See comments for the ARRAY logic. These apply here as well.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,269 @@
+// BufferPtr is modified and results of hash are stored there.
+void PrestoHasher::bulkHash(const SelectivityVector& rows, BufferPtr& buffer) {
+  auto kind = vector_->base()->typeKind();
+  VELOX_DYNAMIC_TYPE_DISPATCH(bulkHash, kind, rows, buffer);
```
bulkHash template is defined separately for each template type, hence, there seems not much benefit from using VELOX_DYNAMIC_TYPE_DISPATCH. A simple switch statement might be clearer.

mbasmanova reviewed (changes requested):
@kgpai I read through the PrestoHasher.h/cpp and VectorMaker.h. Here are some comments.

kgpai commented:
Update: Working on  - 
 - Removing VectorMaker changes to a separate PR.
 - Addressing comments in the PR .
 
These should be ready EOD today.

kgpai commented:
Update: 
 1. Refactored and renamed several variables/functions 
 2. Removed velox type dispatch to more simpler switch
 3. Added more checks for null's and corresponding test cases
 4. Removed vectorMaker array logic to separate PR #841  
 5. Used XXH64 hash for int

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,48 @@
+    createChildren();
+  }
+
+  /// Function that computes a presto equivalent hash.
```
typo: presto -> Presto

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,48 @@
+  /// Function that computes a presto equivalent hash.
+  /// Takes SelectivityVector of rows to be hashed , These rows must be a subset
+  /// of the rows passed to the constructor. The resultant hashes are saved in
+  /// BufferPtr hashes, which must have as much capacity as rows.end().
```
It would be nice to also clarify what is the type of individual hash value so that the caller can allocate the buffer correctly.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,258 @@
+    const DecodedVector& vector,
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  hashFn(rows, hashes, [&](auto row) -> int64_t {
```
"-> int64_t" is redundant

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,258 @@
+
+template <typename Callable>
+FOLLY_ALWAYS_INLINE void
+hashFn(const SelectivityVector& rows, BufferPtr& hashes, Callable func) {
```
Do not abbreviate. I can't think of a good name for this function though. Why not just use rows.applyToSelected directly?

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,258 @@
+      AlignedBuffer::allocate<int64_t>(elementRows.end(), baseArray->pool());
+
+  children_[0]->hash(elementRows, elementHashes);
+  auto indices = vector_->indices();
```
nit: move this variable up and re-use in a call to functions::toElementRows

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,258 @@
+    BufferPtr& hashes) {
+  auto baseRow = vector_->base()->as<RowVector>();
+
+  BufferPtr rowHashBuffer =
```
naming: rowHashBuffer -> childHashes and rowHashes -> rawChildHashes

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,258 @@
+      break;
+    case TypeKind::BOOLEAN:
+      hashFn(rows, hashes, [&](auto row) -> int64_t {
+        return vector_->valueAt<bool>(row) ? 1231 : 1237;
```
null check?

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,258 @@
+      break;
+    case TypeKind::TIMESTAMP:
+      hashFn(rows, hashes, [&](auto row) -> int64_t {
+        return hashInteger((vector_->valueAt<Timestamp>(row)).toMillis());
```
null check?

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,258 @@
+      hash<TypeKind::SMALLINT>(rows, hashes);
+      break;
+    default:
+      VELOX_NYI();
```
use VELOX_UNREACHABLE()

mbasmanova reviewed (commented):
@kgpai I read through the PrestoHasher.h/cpp. Overall looks good, but some null checks might still be missing. Not sure why these are not caught by the tests.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  }
+
+ private:
+  BufferPtr bufferPtr_;
```
used variable? let's remove

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+#include "velox/functions/prestosql/aggregates/tests/AggregationTestBase.h"
+
+namespace facebook::velox::aggregate::test {
+using namespace facebook::velox::aggregate;
```
I think this is not needed.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  }
+
+  template <typename T>
+  void integralTest() {
```
naming: method names start with a verb, e.g. testIntegral

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+
+  template <typename T>
+  void integralTest() {
+    assertHash<T>({1, 0, 100}, {9155312661752487122, 0, -367469560765523433});
```
It would be nice to include a null value.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+class PrestoHasherTest : public AggregationTestBase {
+ protected:
+  template <typename T>
+  void assertHash(
```
It would be easier to read if this method appears next to the other assertHash() method and if the order and names of the parameters were consistent, e.g. data, expected.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+    SelectivityVector rows(vector->size());
+    PrestoHasher hasher(*vector, rows);
+
+    BufferPtr bufferPtr =
```
naming: bufferPtr -> hashes

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  }
+
+  void assertHash(const std::vector<int64_t>& expected, VectorPtr vector) {
+    SelectivityVector rows(vector->size());
```
It would be nice to add some testing for sparsely populated rows.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+        AlignedBuffer::allocate<int64_t>(vector->size(), pool_.get());
+    hasher.hash(rows, bufferPtr);
+
+    auto buffer = bufferPtr->as<int64_t>();
```
naming: rawHashes

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+    hasher.hash(rows, bufferPtr);
+
+    auto buffer = bufferPtr->as<int64_t>();
+    rows.applyToSelected(
```
it would be clearer to use a simple for loop

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+
+    auto buffer = bufferPtr->as<int64_t>();
+    rows.applyToSelected(
+        [&](auto row) { EXPECT_EQ(expected[row], buffer[row]); });
```
nit: consider adding some context to the error message, e.g. `<< "at " << row`

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+        [&](auto row) { EXPECT_EQ(expected[row], buffer[row]); });
+  }
+
+  VectorPtr makeMapofArraysVector(
```
typo: MapOf

mbasmanova replied:
Wouldn't it be easier to use a method that took an std::unordered_map?

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  BufferPtr bufferPtr_;
+};
+
+TEST_F(PrestoHasherTest, integralTypes) {
```
It would be nice to break up this test into multiple tests, one for each type, and make sure to test nulls as well as values at the ends of the supported value range.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+      {Date(0), Date(1000), std::nullopt}, {0, 2343331593029422743, 0});
+}
+
+TEST_F(PrestoHasherTest, doubleType) {
```
nit: doubleType -> doubles; ditto other test names

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+      {-6641611864725600567L, 0, 6018425597037803642, -6781949276808836923});
+}
+
+TEST_F(PrestoHasherTest, VarcharType) {
```
naming: test names start lowercase

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+}
+
+TEST_F(PrestoHasherTest, boolType) {
+  assertHash<bool>({true, false}, {1231, 1237});
```
nulls?

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  };
+
+  auto baseArrayVector = vectorMaker_.arrayVectorNullable<int64_t>(
+      {O({1, 2}),
```
Is the "0" necessary here? I find it hard to read.

kgpai replied:
Unfortunately i dont think c++ compiler is smart enough to auto box vector<optiona<vector<optional<int>>> , so have to use the O's (optionals) - If theres a better way let me know.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+    return std::make_optional(data);
+  };
+
+  auto baseArrayVector = vectorMaker_.arrayVectorNullable<int64_t>(
```
Would be nice to test null and empty arrays as well.

kgpai replied:
Done below..

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+       O({10, 11}),
+       O({12, std::nullopt})});
+
+  auto arrayExpected = std::vector<int64_t>(
```
naming: expectedHashes or expected

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  auto arrayOfArrayVector =
+      vectorMaker_.arrayVector({0, 2, 4}, baseArrayVector);
+
+  auto arrayofArraysExpected = std::vector<int64_t>(
```
typo: arrayOf..., but it would be easier to read if you simply inlined that variable

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+
+  // Array with nulls
+
+  std::vector<std::optional<std::vector<std::optional<int64_t>>>> data = {
```
I'd inline this variable.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+
+  auto arrayWithNulls = vectorMaker_.arrayVectorNullable(data);
+
+  auto arrayWithNullsExpected = std::vector<int64_t>{
```
same here

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+}
+
+TEST_F(PrestoHasherTest, rows) {
+  auto firstVector = makeFlatVector<int64_t>({1, 3});
```
I'd inline these variables for readability.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+
+  auto row = makeRowVector({firstVector, secondVector});
+
+  auto rowExpected =
```
inline

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  auto rowExpected =
+      std::vector<int64_t>{4329740752828761434, 655643799837772474};
+
+  assertHash(rowExpected, row);
```
Would be nice to test nulls.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,261 @@
+  assertHash(rowExpected, row);
+}
+
+TEST_F(PrestoHasherTest, dictionary) {
```
It would be nice to extend assertHash to add testing of encodings, e.g. wrap input in a dictionary, compute hashes, assert result. Then, wrap input in a constant.

kgpai replied:
Done.

mbasmanova reviewed (commented):
@kgpai Some comments on the test.

kgpai commented:
Update: Working on 
 - Adding the dictionary , constant wrapping tests. 
 - Also writing a benchmark to determine performance if nullity checks do not have to take place

kgpai commented:
Update: 
* Added  dictionary , constant wrapping tests and made changes as per comments. 

Moving the benchmark work to #863 .

kgpai commented:
Here are the benchmark results : 

```
============================================================================
/Users/kpai/src/Velox/velox/functions/prestosql/aggregates/benchmarks/DispatchBenchmark.cpprelative  time/iter  iters/s
============================================================================
bulk_dispatch                                               14.88ms    67.21
optimized                                        193.33%     7.70ms   129.93
============================================================================

```

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,285 @@
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(rows, hashes, [&](auto row) {
+    return vector_->isNullAt(row)
```
Null handling is the same for all types. Consider moving it into applyHashFunction so that it is done once. This will also help ensure that null handling is done properly for all types.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,49 @@
+    createChildren();
+  }
+
+  /// Function that computes a Presto equivalent hash.
```
Drop "Function that" for clarity. "Computes ....".

Would readers know what Presto equivalent hash is? Perhaps, clarify , "Computes a hash using algorithm identical to the one used by Presto checksum() aggregate function. This algorithm is based on xxhash64."

mbasmanova commented on velox/functions/prestosql/aggregates/benchmarks/DispatchBenchmark.cpp:
```diff
@@ -0,0 +1,122 @@
+  return hash;
+}
+
+class DispatchBenchmark : public functions::test::FunctionBenchmarkBase {
```
My understanding is that benchmark is still work in progress. To unblock this PR, let's move the benchmark to a separate PR.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+    BufferPtr indices =
+        AlignedBuffer::allocate<vector_size_t>(vectorSize, pool_.get());
+    auto rawIndices = indices->asMutable<vector_size_t>();
+    // Just 1:1 mapping.
```
Let's test something else as well, e.g. a subset of rows as well as repeats of some rows.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+    std::vector<int64_t> constantExpected{0};
+    for (int i = 0; i < vectorSize; i++) {
+      constantExpected[0] = expected[i];
+      auto constantVector = BaseVector::wrapInConstant(1, i, vector);
```
Let's make sure to test vectors of size > 1.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+  testIntegral<int16_t>();
+  testIntegral<int32_t>();
+  testIntegral<int64_t>();
+  assertHash<int32_t>({99999, std::nullopt}, {-6111805494244633133, 0});
```
The code would be more readable if there was a test per type, e.g. tinyints, smallints, integers, etc.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+       3801170566614750614});
+}
+
+TEST_F(PrestoHasherTest, varcharType) {
```
naming consistency: varchars

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+
+TEST_F(PrestoHasherTest, varcharType) {
+  assertHash<StringView>(
+      {StringView("abcd"),
```
nit: "abcd"_sv

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+  assertHash(row, {7113531408683827503, -1169223928725763049});
+}
+
+TEST_F(PrestoHasherTest, dictionary) {
```
This test is no longer needed, right?

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+        {9155312661752487122, 0, -367469560765523433, 0});
+  }
+
+  VectorPtr makeMapOfArraysVector(
```
naming consistency: createMapOfArraysVector

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,343 @@
+template <typename T>
+using limits = std::numeric_limits<T>;
+
+class PrestoHasherTest : public AggregationTestBase {
```
AggregationTestBase is intended to be used for aggregate functions tests. What do you need from this class?

kgpai replied:
I also use a bunch of other make array/map Vector functions; Going to move these utility functions to another base class that can be shared by both PrestoHasher and AggregationTestBase..

mbasmanova reviewed (commented):
@kgpai Looks good overall % small comments below.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,49 @@
+  /// hashed , These rows must be a subset of the rows passed to the
+  /// constructor. The resultant hashes are saved in BufferPtr hashes, which
+  /// must have as much capacity as sizeof(int64_t) * rows.end().
+  void hash(const SelectivityVector& rows, BufferPtr& hashes);
```
A question about the API.

I'm seeing that PrestoHasher instance can be used to compute hashes for a single vector. It can be used to compute multiple sets of hashes for different subset of rows in a given vector. 

Assuming PrestoHasher is intended to be used as a building block for checksum aggregate function, it would be nice to allow for a single instance to be used for hashing multiple vectors of the same type.

PrestoHasher contains DecodedVector's which allocate memory. Reusing PrestoHasher instances would allow for reusing DecodedVector's and the memory they allocate. 

Hence, I envision a slightly different API where PrestoHasher is created using a TypePtr that specified the type of vectors it is capable of hashing and provides a hash() function that takes a vector, a SelectivityVector, and an output buffer of hashes. This way, checksum aggregate function could use a single instance of PrestoHasher to hash all batches of input. 

```
PrestoHasher(TypePtr type);

void hash(const BaseVector& vector, const SelectivityVector& rows, BufferPtr& hashes);
```

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,344 @@
+                           const std::vector<int64_t>& expected) {
+      auto vectorSize = vector->size();
+      SelectivityVector rows(vectorSize);
+      PrestoHasher hasher(*vector, rows);
```
Do we have tests that use a subset of rows? It looks like all tests always specify all rows.

kgpai commented:
Update: 
 * Added repeating and subset of rows dictionary tests
 * Added > 1 constant vector tests
 * Broke up integer tests to smaller tests for smallint,bigint etc..
 * Other naming and misc changes.

kgpai commented:
Going to update the API as per Mashas suggestion and enable Prestohasher to hash for any vector of a particular type. Hoping to get that done by EOD.

kgpai commented:
* Added SelectivityRows subset test in assertHash
* Modified API to take TypePtr in constructor as suggested and vector in hash function
* Refactored PrestoHasherTest to inherit from VectorTestBase class which contains utilities for making vectors.

mbasmanova commented on velox/exec/tests/utils/CMakeLists.txt:
```diff
@@ -21,7 +21,8 @@ add_library(
   OperatorTestBase.cpp
-  HiveConnectorTestBase.cpp)
+  HiveConnectorTestBase.cpp
+  VectorTestBase.h)
```
Do not add header files to CMakeLists.txt.

mbasmanova commented on velox/exec/tests/utils/OperatorTestBase.h:
```diff
@@ -21,12 +21,13 @@
 
 namespace facebook::velox::exec::test {
-class OperatorTestBase : public testing::Test {
+class OperatorTestBase : public testing::Test, public VectorTestBase {
```
Nice. Can we re-use pool_ and vectorMaker_ variables from the base class (VectorTestBase) and remove these from OperatorTestBase?

mbasmanova commented on velox/exec/tests/utils/VectorTestBase.h:
```diff
@@ -0,0 +1,120 @@
+
+namespace facebook::velox::exec::test {
+
+class VectorTestBase {
```
Would it make sense to move this class to velox/vector/tests/ and include it in velox_vector_test_lib library? It would be nice to introduce this class and refactor OperatorTestBase and FunctionTestBase in a separate PR.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,275 @@
+  auto rawHashes = hashes->asMutable<int64_t>();
+
+  rows.applyToSelected([&](auto row) {
+    if (vector.isNullAt(row)) {
```
Thank you for consolidating the null checks in one place. This will make it easier to add an optimization to skip null checks if vector.mayHaveNulls() is false.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,275 @@
+    const VectorPtr& vector,
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  VELOX_USER_CHECK(
```
This should be VELOX_CHECK as this would be a programming error, rather than user error.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,51 @@
+
+  std::shared_ptr<DecodedVector> vector_;
+  std::vector<std::unique_ptr<PrestoHasher>> children_;
+  TypePtr type_;
```
This variable can be const.

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,51 @@
+
+class PrestoHasher {
+ public:
+  explicit PrestoHasher(TypePtr type) : type_(type) {
```
nit: type_(std::move(type))

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,51 @@
+class PrestoHasher {
+ public:
+  explicit PrestoHasher(TypePtr type) : type_(type) {
+    vector_ = std::make_shared<DecodedVector>();
```
nit: this can be initialized when declared, e.g.

```
std::shared_ptr<DecodedVector> vector_{std::make_shared<DecodedVector>()};
```

mbasmanova commented on velox/functions/prestosql/aggregates/PrestoHasher.h:
```diff
@@ -0,0 +1,51 @@
+
+namespace facebook::velox::aggregate {
+
+class PrestoHasher {
```
It would be nice to document this class.

mbasmanova commented on velox/functions/prestosql/aggregates/benchmarks/CMakeLists.txt:
```diff
@@ -11,17 +11,19 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-add_executable(velox_aggregates_pushdown_benchmarks PushdownBenchmark.cpp)
+set(BENCHMARK_DEPENDENCIES
```
These changes are not relevant to this PR, are they?

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+        someRows.setValid(i, true);
+      }
+      someRows.updateBounds();
+      PrestoHasher someHash(vector->type());
```
It would be nice to reuse the "hasher" instance to provide more coverage for the common use case of re-using the PrestoHasher instances.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+      }
+      someRows.updateBounds();
+      PrestoHasher someHash(vector->type());
+      BufferPtr someHashes =
```
It would be nice to re-use hashes buffer from above to provide more coverage for the common use case.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+      BufferPtr someHashes =
+          AlignedBuffer::allocate<int64_t>(someRows.end(), pool_.get());
+      someHash.hash(vector, someRows, someHashes);
+      auto someRawHashes = someHashes->as<int64_t>();
```
naming: rawSomeHashes

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+      }
+
+      // Check for a subset of rows.
+      SelectivityVector someRows(vectorSize, false);
```
nit: consider extracting helper method to create SelectivityVector for even rows, e.g. 

```
auto evenRows = selectEvenRows(vectorSize);
```

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+    checkHashes(vector, expected);
+
+    // Wrap in Dictionary
+    BufferPtr indices =
```
Consider using makeIndices function from the base class:

`makeIndices(vectorSize, [] (auto row) {return row;})`

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+
+    // Subset of rows.
+    auto subsetSize = vectorSize / 2;
+    BufferPtr subsetIndices =
```
Any reason this code can't reuse indices buffer?

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+    BufferPtr subsetIndices =
+        AlignedBuffer::allocate<vector_size_t>(subsetSize, pool_.get());
+    auto rawSubsetIndices = subsetIndices->asMutable<vector_size_t>();
+    std::vector<int64_t> subsetExpected;
```
Any reason 'modifiedExpected' variable can't be reused?

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+    std::vector<int64_t> subsetExpected;
+    subsetExpected.reserve(vectorSize / 2);
+
+    for (size_t i = 0; i < vectorSize / 2; ++i) {
```
vectorSize / 2 -> subsetSize

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+          std::optional<std::vector<std::optional<int64_t>>>> map) {
+    std::vector<std::optional<int64_t>> keys;
+    std::vector<std::optional<std::vector<std::optional<int64_t>>>> values;
+    for (const auto& [key, vals] : map) {
```
vals -> value

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+      values.push_back(vals);
+    }
+
+    auto arrayValues = vectorMaker_.arrayVectorNullable(values);
```
Is there no corresponding makeNullableArrayVector in the base class? Let's add one.

mbasmanova replied:
arrayValues -> mapValues for consistency

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,371 @@
+  assertHash(row, {7113531408683827503, -1169223928725763049});
+}
+
+TEST_F(PrestoHasherTest, differentVector) {
```
nit: differentVector -> wrongVectorType or wrongType

mbasmanova reviewed (commented):
@kgpai Looks great. A few remaining comments below.

kgpai commented:
* Removed VectorBaseTest to a separate PR 
* Refactoring tests as per comments.
* Misc variable name changes as per PR comments.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp:
```diff
@@ -0,0 +1,361 @@
+
+    // Wrap in Dictionary
+    BufferPtr indices = makeIndices(vectorSize, [](auto row) { return row; });
+    auto dictionaryVector = BaseVector::wrapInDictionary(
```
nit: you can now use wrapInDictionary helper method from the base class

```
wrapInDictionary(indices, vectorSize, vector);
```

mbasmanova reviewed (approved):
@kgpai Thank you. Looks good to me.
